src/OpenApi/src/OpenApiGenerator.cs

```diff
@@ -0,0 +1,474 @@
+                }
+            }
+
+            var nullabilityContext = new NullabilityInfoContext();
+            var nullability = nullabilityContext.Create(requestBodyParameter);
+            var allowEmpty = requestBodyParameter.GetCustomAttributes().OfType<IFromBodyMetadata>().SingleOrDefault()?.AllowEmpty ?? false;
+            var isOptional = requestBodyParameter.HasDefaultValue
+                || nullability.ReadState != NullabilityState.NotNull
+                || allowEmpty;
+
+            return new OpenApiRequestBody
+            {
+                Required = !isOptional,
+                Content = requestBodyContent
+            };
+        }
+
+        return null;
+    }
+
+    private List<OpenApiTag> GetOperationTags(MethodInfo methodInfo, EndpointMetadataCollection metadata)
+    {
+        var tags = metadata.GetMetadata<ITagsMetadata>();
+        string controllerName;
+
+        if (methodInfo.DeclaringType is not null && !TypeHelper.IsCompilerGeneratedType(methodInfo.DeclaringType))
+        {
+            controllerName = methodInfo.DeclaringType.Name;
+        }
+        else
+        {
+            // If the declaring type is null or compiler-generated (e.g. lambdas),
+            // group the methods under the application name.
+            controllerName = _environment?.ApplicationName ?? string.Empty;
+        }
+
+        return tags is not null
+            ? tags.Tags.Select(tag => new OpenApiTag() { Name = tag }).ToList()
+            : new List<OpenApiTag>() { new OpenApiTag() { Name = controllerName } };
+    }
+
+    private List<OpenApiParameter> GetOpenApiParameters(MethodInfo methodInfo, EndpointMetadataCollection metadata, RoutePattern pattern, bool disableInferredBody)
+    {
+        var parameters = methodInfo.GetParameters();
+        var openApiParameters = new List<OpenApiParameter>();
+
+        foreach (var parameter in parameters)
+        {
+            var (isBodyOrFormParameter, parameterLocation) = GetOpenApiParameterLocation(parameter, pattern, disableInferredBody);
+
+            // If the parameter isn't something that would be populated in RequestBody
+            // or doesn't have a valid ParameterLocation, then it must be a service
+            // parameter that we can ignore.
+            if (!isBodyOrFormParameter && parameterLocation is null)
+            {
+                continue;
+            }
+
+            var nullabilityContext = new NullabilityInfoContext();
+            var nullability = nullabilityContext.Create(parameter);
+            var isOptional = parameter.HasDefaultValue || nullability.ReadState != NullabilityState.NotNull;
+            var openApiParameter = new OpenApiParameter()
+            {
+                Name = parameter.Name,
+                In = parameterLocation,
+                Content = GetOpenApiParameterContent(metadata),
+                Schema = new OpenApiSchema { Type = SchemaGenerator.GetOpenApiSchemaType(parameter.ParameterType) },
```

Comment:
This should take TryParse into account, right?
https://github.com/dotnet/aspnetcore/blob/64ccf823287a1de02385d29497127ad5197a389d/src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs#L284-L286

Reply:
The rule has to factor in the underlying type of the type with the `TryParse` so the check is a little more involved with that. Added a callout to it in https://github.com/dotnet/aspnetcore/issues/41246